### PR TITLE
`key_share`: add `no_std` support

### PR DESCRIPTION
**Pull request by survived**
blocked by https://github.com/dfns/generic-ec/pull/25

### CODE REVIEW

**Review by drewstone — APPROVED** [review]

**Review by maurges — DISMISSED** [review]

**Comment by maurges** [comment]
Instead of making Display instances by hand, we could convert what we had with `thiserror` into https://docs.rs/displaydoc/

**Comment by survived** [comment]
Nice! I'll try using it. Tho I'm not really a fan of deriving display impl from doc comments

**Comment by survived** [comment]
I see that I can use `#[displaydoc("...")]` instead. I'll give it a shot

**Comment by survived** [comment]
@maurges I changed the code to use `displaydoc` + `thiserror`. A bit prettier, but it did introduce a breaking change: the `HdError<E>` now only implements `Display` if `E: Display` (it used to implement display for regardless of `E`). However, I doubt it's an any significant breaking change...

File: key-share/src/lib.rs [hunk 1]
```diff
@@ -468,41 +475,51 @@ impl<E: Curve> AsRef<CoreKeyShare<E>> for CoreKeyShare<E> {
 }
 
 /// Error indicating that key share is not valid
-#[derive(Debug, thiserror::Error)]
-#[error(transparent)]
-pub struct InvalidCoreShare(#[from] InvalidShareReason);
+#[derive(Debug, displaydoc::Display)]
+#[cfg_attr(feature = "std", derive(thiserror::Error))]
+#[displaydoc("invalid core key share")]
+pub struct InvalidCoreShare(#[cfg_attr(feature = "std", source)] InvalidShareReason);
```
Comment (maurges) [inline key-share/src/lib.rs:481]: Why change from transparent?

Comment (maurges) [reply key-share/src/lib.rs:481]: I guess because a transparent error implies an impossible Display implementation?

**Comment by maurges** [comment]
> the `HdError<E>` now only implements `Display` if `E: Display`

It's silly that displaydoc crate doesn't allow us to request bounds by hand. But I agree that this is not really and important change

**Review by maurges — APPROVED** [review]

File: key-share/src/lib.rs (hunk 1 above)
Comment (survived) [reply key-share/src/lib.rs:481]: Transparent error will generate Display implementation no matter what. We want `displaydoc` to do Display impl, not `thiserror`

**Pull request merged**